### PR TITLE
Nuclear Operative Commanders now start with a Commander Uplink

### PR DESCRIPTION
**Pull request by Zonespace27**
[Balance] [Feature]
## About the PR 
This PR adds a commander uplink for Nuclear Operative Commanders, giving them the ability to get special equipment to level the playing field for Nuclear Operatives.
Currently, the uplink doesn't use traditional TC, instead using "Points". The uplink gets 1 point for every 20 people on the server, with a minimum of 2 points.
The equipment that the uplink has goes as follows:
**Note that everything in the clips are older, so things may have changed from what's seen in the video.**

- [Reinforcements (Video OOD)](https://imgur.com/a/DQmG9qy) (2 Points): This gives a beacon, which on use, polls the ghost pool for one ghost to become a "Syndicate Reinforcement". Once picked, they will be deployed to the beacon's location (only usable on z1) via an arrivals missile. This reinforcement is a Syndicate retrofitted gunbot, being tankier, in addition to having their .38 gun replaced with a 7.62 DMR.

- [Ammo Bag](https://imgur.com/a/uUoKzkI) (2 Points): This one gives the commander a folded-up ammo bag, which can be deployed anywhere with an action bar. Once deployed, anyone can hit the bag with a Syndicate weapon to get its respective ammo, using a certain number of "charges" (bag starts with 10) depending on the size of the ammo created.

- [Specialist Ammo Bag](https://imgur.com/a/uUoKzkI) (3 Points): This bag is similar to the standard ammo bag, except that it allows the creation of "specialist" ammo, usually being AP rounds (in the case of the Predator and Sirius), but can be other types as well (Orion gets HP rounds, SPES-6 gets normal buckshot). In addition, it can fabricate ammo the normal ammo bag can't, notably being the grenadier's 40mm grenades and the MRPT's rockets. If you try to fabricate ammo for a gun with no special types (for example, the Branwen), the ammo will cost one less "charge" to fabricate versus the regular ammo bag.

- [Laser Designator](https://imgur.com/a/p0Kgnhq) (3 Points): This gives the commander a monocular laser designator (Thank you Waffleloffle for the sprite!) that allows the commander to scope in (like the Betelgeuse Sniper) and designate a target at range. Once a target is designated, the tile gains an overlay, in addition to a station announcement being made of what area the artillery will land in. Within 10-15 seconds, the artillery will land, causing a 3x3 tile destruction (75 power). In addition, the laser designator is fully coded to be flexible to add, and requires the newly-added 152mm cannon on the Cairngorm to be working and loaded (currently impossible to destroy and has infinite ammo, but it does check!). The designator has two charges.

- [Rapid Team Deployment Remote](https://imgur.com/a/0lAlJw9) (4 Points): Finally, this gives the commander a handheld remote. This remote allows the commander, along with the nuke and anyone in range (4 tiles) to deploy through arrival missiles to any area on-station. How it works is that first, the commander chooses an area to deploy to. Upon using the remote in hand again, it asks the commander to confirm that they are ready to deploy, along with listing how many team members are in range, and if the nuke is as well. Once the commander has extra-confirmed that they are ready to go, everyone in range, along with the nuke, will be teleported to the "Syndicate Assault pod". Once the team is in there, a visible timer (which is about a minute long) will start counting down until the team gets launched in via arrival missiles into the pre-specified area. During the time in the assault pod, the station will get three announcements total, the first warning the crew of the nuke team's existence, the second saying the touchdown location, and the third being the amount of nukies along with the touchdown location.

- Nuclear Bomb Teleporter Remote (1 Point): Moved from the commander's starting gear to uplink. Otherwise, functions like normal.

## Why's this needed? 
It was on [Gannet's](https://github.com/goonstation/goonstation/projects/8#card-66441395) [wishlist](https://github.com/goonstation/goonstation/projects/8#card-66441377), and I thought the idea was quite interesting!

## TODO:

- [x] Finish changelog
- [x] Get ammobag sprites
- [x] Possibly get a unique gunbot sprite?
- [x] Wait for Cogwerks to finish the 152mm sprite & review placement of the 152mm gun

## Changelog

```changelog
(u)Zonespace
(*)Nuclear Operative Commanders now have a specialized uplink! More info in Minor Changes. Nuclear bomb remote has been moved to the uplink.
(+)The commander's uplink uses "points" versus the traditional TC, and gains 1 point for every 20 people online (minimum of 2 points)
(+)The uplink has 6 items, being the Nuclear bomb teleporter, reinforcement beacon, ammo bag, specialist ammo bag, laser designator, and rapid deployment remote, the new ones described in further changelog entries.
(+)First is a reinforcement beacon for 2 points to summon a syndicate gunbot to assist the nuclear team.
(+)Second is the Ammo Bag and its specialist version (2/3 points, respectively), which both allow the nuclear operatives to fabricate more ammunition for their guns.
(+)Third is the laser designator for 3 points, which allows the commander to call in two high-explosive airstrikes from the Cairngorm to hit anywhere they designate!
(+)Finally is the Rapid Deployment Remote, costing a whole 4 points, which allows the commander to send the nuclear team in through drop pods onto anywhere on station!
(+)Thank you to Cogwerks for the 152mm gun sprites, Waffleoffle for the laser designator/ammobag sprites, and Dimwhat for the gunbot/drop pod sprites!
```

### CODE REVIEW

**Comment by Zonespace27** [comment]
> Wondering what Gannets thinks, but I feel like the bomb remote could be moved to this. IIRC it was added when crew could still steal the nuke, so it's lost the easy plant vs retrieving a stolen nuke thing and is basically just a free plant now.

Good idea, added.

**Comment by Zonespace27** [comment]
Looked through the target designator projectile, it's not really what i'm looking for with the laser designator tbh

**Comment by Crystalwarrior** [comment]
Hey, why this change?

> Nuclear Bomb Teleporter Remote (1 Point): Moved from the commander's starting gear to uplink. Otherwise, functions like normal.

This makes the up link have zero effect on solo nuke op as the minimum is 2 points, and the teleporter is the only 1 point item. Isn't that a bit needless?

**Comment by Zonespace27** [comment]
It was suggested or liked by multiple people (like Gannets), it reduces the potential bloat of what the commander starts with, and it fits well in uplink

**Comment by Zonespace27** [comment]
Small update: uplink is now tgui-ified, here's a slightly out-of-date photo.
image

**Comment by Zonespace27** [comment]
A big thanks to Dimwhat for the gunbot sprite!
image

**Comment by Zonespace27** [comment]
image
Thanks to Waffleoffle for the ammobag sprites!

**Review by Gannets — APPROVED** [review]
I haven't looked at the code, but I'm drive-by approving before my in-laws get here for boxing day. let's do it and see what happens!

File: code/obj/item/uplinks.dm [hunk 1]
```diff
@@ -952,7 +952,169 @@ Note: Add new traitor items to syndicate_buylist.dm, not here.
...
+			))
+
+	ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+		. = ..()
+		if(.)
+			return
+		switch(action)
+			if ("redeem")
+				for(var/datum/syndicate_buylist/SB as anything in commander_buylist)
+					if(istype(commander_buylist[SB], locate(params["ref"])) || commander_buylist[SB] == params["ref"]) //completely untested shitcode, look at later
```
Comment (ZeWaka) [inline code/obj/item/uplinks.dm:1054]: ?

File: tgui/packages/tgui/styles/interfaces/ComUplink.scss [hunk 2]
```diff
@@ -0,0 +1,28 @@
+@use '../colors.scss';
```
Comment (ZeWaka) [inline tgui/packages/tgui/styles/interfaces/ComUplink.scss:1]: License

File: code/obj/item/uplinks.dm [hunk 3]
```diff
@@ -952,7 +952,169 @@ Note: Add new traitor items to syndicate_buylist.dm, not here.
...
+							if(B.item2)
+								new B.item2(get_turf(src))
+							if(B.item3)
+								new B.item3(get_turf(src))
+							B.run_on_spawn(A, usr)
+							. = TRUE
+							break
+
+#define CHECK1 (get_dist(src, usr) > 1 || !usr.contents.Find(src) || !isliving(usr) || iswraith(usr) || isintangible(usr))
+#define CHECK2 (is_incapacitated(usr) || usr.restrained())
```
Comment (ZeWaka) [inline code/obj/item/uplinks.dm:1068]: Poor names

File: code/obj/item/device/nuclear_op_items.dm [hunk 4]
```diff
@@ -0,0 +1,919 @@
...
+		var/list/text_messages = list()
+		text_messages.Add("Would you like to respawn as a nuclear operative gunbot? You may be randomly selected from the list of candidates.")
+		text_messages.Add("You are eligible to be respawned as a nuclear operative gunbot. You have [src.ghost_confirmation_delay / 10] seconds to respond to the offer.")
+		text_messages.Add("You have been added to the list of eligible candidates. Please wait for the game to choose, good luck!")
+
+		// The proc takes care of all the necessary work (job-banned etc checks, confirmation delay).
+		message_admins("Sending Syndicate Reinforcement offer to eligible ghosts. They have [src.ghost_confirmation_delay / 10] seconds to respond.")
+		var/list/datum/mind/candidates = dead_player_list(1, src.ghost_confirmation_delay, text_messages, allow_dead_antags = 1)
+		if(!length(candidates))
+			src.visible_message("<span class='alert'>The [src] buzzes, before unbolting itself from the ground. There seems to be no reinforcements available currently.</span>")
```
Comment (ZeWaka) [inline code/obj/item/device/nuclear_op_items.dm:91]: Would suck on lowpop

File: code/obj/item/device/nuclear_op_items.dm [hunk 5]
```diff
@@ -0,0 +1,919 @@
...
+				var/obj/decal/teleport_swirl/S = new/obj/decal/teleport_swirl(NB)
+				NB.set_loc(L)
+				var/obj/decal/residual_energy/R = new/obj/decal/residual_energy(L)
+				playsound(L, "sound/effects/teleport.ogg", 30, 1)
+				SPAWN_DBG(1 SECOND)
+					qdel(S)
+					qdel(R)
+			nuclear_bombs += NB
+		src.used = TRUE
+		for(var/obj/machinery/computer/security/pod_timer/S in range(1, pick_landmark(LANDMARK_SYNDICATE_ASSAULT_POD_COMP))) //This is the only way I could make this work
```
Comment (ZeWaka) [inline code/obj/item/device/nuclear_op_items.dm:415]: what? this doesn't make sense

File: code/obj/item/device/nuclear_op_items.dm [hunk 6]
```diff
@@ -0,0 +1,919 @@
...
+	bound_width = 96
+	firingfrom = ""
+	ammo = -1
+	sound_offset_dir = EAST
+	sound_offset_length = 3
+
+	bombard(atom/target, mob/user)
+		var/turf/target_turf = get_turf(target)
+		var/turf/firing_turf = get_turf(src)
+		if(!(target_turf in view(DESIGNATOR_MAX_RANGE, usr.loc))) //view() is bad and slow but I cannot find a better way to do this
```
Comment (ZeWaka) [inline code/obj/item/device/nuclear_op_items.dm:665]: what are you trying to do

File: code/obj/item/device/nuclear_op_items.dm [hunk 7]
```diff
@@ -0,0 +1,919 @@
...
+	name = "Generic Ammobox"
+	desc = "You shouldn't see me!"
+	icon_state = "lmg_ammo-0-old"
+	ammo_type = null
+	caliber = null
+	var/list/valid_calibers = list() //supports lists and single, set to "All" for any gun
+
+	attackby(obj/item/gun/kinetic/W, mob/user)
+		if(istype(W, /obj/item/gun/kinetic))
+			if((islist(valid_calibers) && valid_calibers.Find(initial(W.caliber))) || (!islist(valid_calibers) && valid_calibers == initial(W.caliber)))
```
Comment (ZeWaka) [inline code/obj/item/device/nuclear_op_items.dm:713]: you don't need .Find, you're not using the index returned

File: code/obj/item/device/nuclear_op_items.dm [hunk 8]
```diff
@@ -0,0 +1,919 @@
...
+				else
+					boutput(user, "<span class='alert'>The [src] doesn't have enough charge left to fabricate the ammo for [W]!</span>")
+					return
+
+			else if(istype(W, /obj/item/gun/kinetic/silenced_22))
+				if(src.charge >= 1)
+					var/obj/O = W.default_magazine
+					if(spec_ammo)
+						new O(get_turf(src))
+						boutput(user, "<span class='alert'>You get a [initial(O.name)] out of [src].</span>")
```
Comment (ZeWaka) [inline code/obj/item/device/nuclear_op_items.dm:835]: endless copypasta gross

File: code/obj/item/device/nuclear_op_items.dm [hunk 9]
```diff
@@ -0,0 +1,919 @@
...
+
+			// Guns with special ammo types above //
+			else if(istype(W, /obj/item/gun/kinetic/pistol))
+				if(src.charge >= 1)
+					var/obj/O = W.default_magazine
+					new O(get_turf(src))
+					boutput(user, "<span class='alert'>You get a [initial(O.name)] out of [src].</span>")
+					src.charge -= 1
+				else
+					boutput(user, "<span class='alert'>The [src] doesn't have enough charge left to fabricate the ammo for [W]!</span>")
```
Comment (ZeWaka) [inline code/obj/item/device/nuclear_op_items.dm:853]: plox

**Review by ZeWaka — CHANGES_REQUESTED** [review]
I also didn't see an indicator for the laser at the position

**Comment by ZeWaka** [comment]
If you do work like this in the future, you will need to split it up into multiple PRs. There's a reason there hasn't been a comprehensive review on this one.

File: code/obj/item/device/nuclear_op_items.dm (hunk 6 above)
Comment (Zonespace27) [reply code/obj/item/device/nuclear_op_items.dm:665]: Preventing from designating a target from behind an opaque wall provided you can't see it, I've been told `view()` is generally a bit slow

Comment (ZeWaka) [reply code/obj/item/device/nuclear_op_items.dm:665]: ask pali or someone about a proc for this, it's based on the resenham line drawing algorithim, it's like can_see or something - or maybe `viewers` (unsure if works for non-mobs)

File: code/obj/item/device/nuclear_op_items.dm (hunk 4 above)
Comment (Zonespace27) [reply code/obj/item/device/nuclear_op_items.dm:91]: I don't really know how to help with this, you can already refund it by hitting the uplink with it, and nukeops can just kill people to get more ghosts

File: code/obj/item/device/nuclear_op_items.dm (hunk 5 above)
Comment (Zonespace27) [reply code/obj/item/device/nuclear_op_items.dm:415]: I tried creating a `new` computer on the landmark and editing its maptext, but that didn't work, I eventually landed on this

File: code/obj/item/device/nuclear_op_items.dm (hunk 4 above)
Comment (ZeWaka) [reply code/obj/item/device/nuclear_op_items.dm:91]: Fair enough.

**Comment by ZeWaka** [comment]
Why mark my comments resolved without addressing them in code?

**Comment by Zonespace27** [comment]
> Why mark my comments resolved without addressing them in code?

I have uncommitted code locally that addresses them, I just haven't committed it yet

**Comment by Zonespace27** [comment]
@ZeWaka Made requested changes

**Comment by ZeWaka** [comment]
The .Find replacement will runtime, you should've replaced with `xyz in list_var`

File: code/obj/item/device/nuclear_op_items.dm [hunk 10]
```diff
@@ -0,0 +1,824 @@
...
+	name = "Generic Ammobox"
+	desc = "A generic ammobox for getting some ammunition."
+	icon_state = "lmg_ammo-0-old"
+	ammo_type = null
+	caliber = null
+	var/list/valid_calibers = list() //supports lists and single, set to "All" for any gun
+
+	attackby(obj/item/gun/kinetic/W, mob/user)
+		if(istype(W, /obj/item/gun/kinetic))
+			if((islist(valid_calibers) && valid_calibers[initial(W.caliber)]) || (!islist(valid_calibers) && valid_calibers == initial(W.caliber)) || valid_calibers == "All")
```
Comment (Zonespace27) [inline code/obj/item/device/nuclear_op_items.dm:714]:
```suggestion
			if((islist(valid_calibers) && initial(W.caliber) in valid_calibers || (!islist(valid_calibers) && valid_calibers == initial(W.caliber)) || valid_calibers == "All")
```

File: tgui/packages/tgui/interfaces/ComUplink/index.tsx [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+ * @license MIT
+ */
+
+import { toTitleCase } from 'common/string';
+import { useBackend } from '../../backend';
+import { Box, Button, Collapsible, LabeledList, Section, Stack, Table } from '../../components';
+import { Window } from '../../layouts';
+import { ComUplinkData, ComUplinkStockData } from './type';
+
+export const pluralize = (word, n) => (n !== 1 ? word + 's' : word);
```
Comment (ZeWaka) [inline tgui/packages/tgui/interfaces/ComUplink/index.tsx:14]: should be in a helper library already if not move it there

File: code/obj/item/gun/kinetic.dm [hunk 12]
```diff
@@ -13,6 +13,9 @@ ABSTRACT_TYPE(/obj/item/gun/kinetic)
 	var/auto_eject = 0 // Do we eject casings on firing, or on reload?
 	var/casings_to_eject = 0 // If we don't automatically ejected them, we need to keep track (Convair880).
 	var/default_magazine //What's the default magazine used in this gun? Set this in place of putting the type in New()
+	var/list/ammobag_magazines = list() //Assoc list of magazine types, standard ammo first, special ammo second
+	var/ammobag_spec_required = FALSE //Can only special-ammo ammobags restock these?
+	var/ammobag_restock_cost = 0 //How many charges it costs an ammobag to fabricate ammo for this gun
```
Comment (ZeWaka) [inline code/obj/item/gun/kinetic.dm:18]: should really be doc comments but whatever

File: code/obj/item/device/nuclear_op_items.dm [hunk 13]
```diff
@@ -0,0 +1,824 @@
...
+		anchored = TRUE
+		deployed = TRUE
+		icon_state = "[initial(icon_state)]-d[charge <= 0 ? "-empty" : ""]"
+
+	proc/fold_ammobag(var/mob/user)
+		anchored = FALSE
+		deployed = FALSE
+		icon_state = "[initial(icon_state)]"
+		sleep(1 DECI SECOND)
+		Attackhand(user)
```
Comment (ZeWaka) [inline code/obj/item/device/nuclear_op_items.dm:768]: what is this

File: code/obj/item/device/nuclear_op_items.dm [hunk 14]
```diff
@@ -0,0 +1,824 @@
...
+	bombard(atom/target, mob/user)
+		var/turf/target_turf = get_turf(target)
+		var/turf/firing_turf = get_turf(src)
+		if(getlineopaqueblocked(get_turf(user), target_turf) != target_turf)
+			return FALSE
+		if(!isnull(src.target_overlay))
+			target_turf.overlays += src.target_overlay
+		while(sound_offset_length > 0)
+			sound_turf = get_step(src, sound_offset_dir)
+			sound_offset_length--
```
Comment (ZeWaka) [inline code/obj/item/device/nuclear_op_items.dm:672]: this looks weirdly useless

File: code/obj/item/device/nuclear_op_items.dm [hunk 15]
```diff
@@ -0,0 +1,824 @@
...
+
+		just_stop_designating(M)
+
+	proc/just_stop_designating(var/mob/living/M) // remove overlay here
+		if (M.client)
+			M.client.pixel_x = 0
+			M.client.pixel_y = 0
+
+		M.use_movement_controller = null
+		M.keys_changed(0,0xFFFF)
```
Comment (ZeWaka) [inline code/obj/item/device/nuclear_op_items.dm:542]: need a comment on this

File: code/obj/item/device/nuclear_op_items.dm (hunk 5 above)
Comment (ZeWaka) [reply code/obj/item/device/nuclear_op_items.dm:415]: no like `range(1 to THING)` makes no sense - does it use its loc or something weird??

File: code/obj/item/device/nuclear_op_items.dm [hunk 16]
```diff
@@ -0,0 +1,824 @@
...
+			SPAWN_DBG(0)
+				var/L = pick_landmark(LANDMARK_SYNDICATE_ASSAULT_POD_TELE)
+				if(!L) //fuck
+					return
+				for(var/obj/item/remote/syndicate_teleporter/T in M.get_all_items_on_mob())
+					qdel(T) //Emphasizing that there really is no easy way back if you go this way
+				playsound(M, "sound/effects/teleport.ogg", 30, 1)
+				var/obj/decal/teleport_swirl/S = new/obj/decal/teleport_swirl(M)
+				M.set_loc(L)
+				var/obj/decal/residual_energy/R = new/obj/decal/residual_energy(L)
```
Comment (ZeWaka) [inline code/obj/item/device/nuclear_op_items.dm:394]: please stop with single letter vars for non-core types

especially if they conflict. T is normally turf. i would not think it's a teleporter.

File: code/obj/item/device/nuclear_op_items.dm [hunk 17]
```diff
@@ -0,0 +1,824 @@
...
+			sent_mobs += M
+		for(var/obj/machinery/nuclearbomb/NB in range(4, user.loc))
+			SPAWN_DBG(0)
+				var/L = pick_landmark(LANDMARK_SYNDICATE_ASSAULT_POD_TELE)
+				if(!L)
+					return
+				playsound(NB, "sound/effects/teleport.ogg", 30, 1)
+				var/obj/decal/teleport_swirl/S = new/obj/decal/teleport_swirl(NB)
+				NB.set_loc(L)
+				var/obj/decal/residual_energy/R = new/obj/decal/residual_energy(L)
```
Comment (ZeWaka) [inline code/obj/item/device/nuclear_op_items.dm:408]: ditto

File: code/obj/item/device/nuclear_op_items.dm [hunk 18]
```diff
@@ -0,0 +1,824 @@
+////////// Nuclear Bomb Teleporter //////////
```
Comment (ZeWaka) [inline code/obj/item/device/nuclear_op_items.dm:1]: this file should really be split up into multiple.

Comment (ZeWaka) [reply code/obj/item/device/nuclear_op_items.dm:1]: also i would never look in obj/item/device for any of this

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: code/obj/item/device/nuclear_op_items.dm (hunk 14 above)
Comment (Zonespace27) [reply code/obj/item/device/nuclear_op_items.dm:672]: The gun faces left, and so the sound will always be on the end of the barrel, which is in space, which means people nearby can't hear it.

File: code/obj/item/device/nuclear_op_items.dm (hunk 13 above)
Comment (Zonespace27) [reply code/obj/item/device/nuclear_op_items.dm:768]: A way to automatically pick it up after folding it up

File: code/obj/item/device/nuclear_op_items.dm (hunk 5 above)
Comment (Zonespace27) [reply code/obj/item/device/nuclear_op_items.dm:415]: `range` takes two arguments, distance and starting location, it checks all turfs within 1 tile (so, on the tile) for a pod_timer computer

File: code/obj/item/device/nuclear_op_items.dm (hunk 15 above)
Comment (Zonespace27) [reply code/obj/item/device/nuclear_op_items.dm]: I don't know _exactly_ how this works, it was just code inherited from the sniper's code, but it breaks without it (at least in my testing)

File: code/obj/item/device/nuclear_op_items.dm (hunk 5 above)
Comment (ZeWaka) [reply code/obj/item/device/nuclear_op_items.dm]: fuck i was thinking in rust

**Comment by Zonespace27** [comment]
Will fix merge conflicts when a dev pings me that it's good to merge

File: _std/__build.dm [hunk 19]
```diff
@@ -97,7 +97,7 @@ o+`        `-` ``..-:yooos-..----------..`
 //#define MAP_OVERRIDE_CHIRON			// Chiron by Kusibu
 //#define MAP_OVERRIDE_OSHAN			// Oshan
 //#define MAP_OVERRIDE_HORIZON			// Horizon by Warcrimes
-//#define MAP_OVERRIDE_ATLAS			// gannetmap OR IS IT KUBIUSGANNETMAP??
+#define MAP_OVERRIDE_ATLAS			// gannetmap OR IS IT KUBIUSGANNETMAP??
```
Comment (Zonespace27) [inline _std/__build.dm:100]:
```suggestion
//#define MAP_OVERRIDE_ATLAS			// gannetmap OR IS IT KUBIUSGANNETMAP??
```

**Comment by ZeWaka** [comment]
why does the changelog say 6 items when you describe 4

**Comment by Zonespace27** [comment]
image
All 6 were described, but CL has been updated with better wording

**Pull request merged**